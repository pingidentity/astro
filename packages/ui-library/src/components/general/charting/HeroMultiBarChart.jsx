import React, { Component, useEffect } from "react";
import PropTypes from "prop-types";
import _ from "lodash";
import {
    ResponsiveContainer,
    BarChart,
    XAxis,
    YAxis,
    CartesianGrid,
    Bar,
    Cell,
    LabelList,
    Tooltip } from "recharts";
import ChartTitle from "./ChartTitle";
import PageSpinner from "../PageSpinner";
import Icon, { iconSizes } from "../Icon";
import Legend, {
    alignments as legendAlignments,
    boxAlignments,
} from "./Legend";
import Padding from "../../layout/Padding";

const SectionHoverHandler = ({ onChange, ...rest }) => {
    useEffect(() => {
        onChange(rest);
    },[rest.label, rest.active]);

    return null;
};

/**
* @class HeroMultiBarChart
* @desc A charting component that renders a single full-width bar chart. This chart contains multi bar chart.
*
* @param {string} [data-id="hero-multi-bar-chart"]
*     The data-id assigned to the top-most container of the component.
* @param {string} [bgImage]
*     A url to an image to display in the background of the component.
* @param {number} [chartWidth=800]
*     The width of the area containing the bar graph
* @param {number} [chartHeight=250]
*     The text renderered in the upper-left of the component
* @param {boolean} [loading=false]
*     When true the splinner animation shows in place of the charts
* @param {string} [loadingMessage="Loading..."]
*     The optional text to display below the loading spinner
* @param {string} [errorMessage]
*     When provided, the error message and icon will display in place of the chart and center text.
* @param {string} [xAxisKey="id"]
*    The object property that contains the x-axis data in the "data" prop
* @param {string} [title]
*    The value for chart title
* @param {string} [array]
*     The optional an array of objects containing the data for Legend component
* @param {array} data
*     An array of objects containing the data for multi series bar chart
* @param {aray} dataKeys
*    An object containing the keys for y-axis data in the "data" prop .
* @param {object} rockerButtonProps
*     An object containing the props passed to the range-selector RockerButton component
* @param {function} [labelFormater]
*     An optional prop with formater for tooltip label
* @param {function} [onBarMouseOver]
*     An optional prop with handler for bar mouse over
* @param {function} [onBarMouseOut]
*     An optional prop with handler for bar mouse out
*/

const labelHeight = 20;

export default class HeroMultiBarChart extends Component {

    static propTypes = {
        "data-id": PropTypes.string,
        bgImage: PropTypes.string,
        chartWidth: PropTypes.number,
        chartHeight: PropTypes.number,
        errorMessage: PropTypes.string,
        loading: PropTypes.bool,
        loadingMessage: PropTypes.string,
        title: PropTypes.string,
        legend: PropTypes.array,
        xAxisKey: PropTypes.string,
        data: PropTypes.array,
        dataKeys: PropTypes.array.isRequired,
        rockerButtonProps: PropTypes.object,
        onBarMouseOver: PropTypes.func,
        onBarMouseOut: PropTypes.func,
        onGroupSelectionChange: PropTypes.func,
    };

    static defaultProps = {
        "data-id": "hero-multi-bar-chart",
        bgImage: "",
        chartWidth: 800,
        chartHeight: 250,
        loading: false,
        loadingMessage: "Loading...",
        xAxisKey: "id",
        onBarMouseOver: _.noop,
        onBarMouseOut: _.noop,
        onGroupSelectionChange: _.noop,
    };

    state = {
        barSelected: null
    };

    _handleBarMouseOver = (key, index, entry) => () => {
        this.props.onBarMouseOver(key, index, entry);
        this.setState({ barSelected: { key, index, entry } });
    }

    _handleBarMouseOut = () => {
        this.props.onBarMouseOut();
        this.setState({ barSelected: null });
    }

    _kFormatter = (value) => Math.abs(value) > 999
        ? Math.sign(value) * ((Math.abs(value) / 1000).toFixed(1)) + "K"
        : Math.sign(value) * Math.abs(value);

    _renderXAxis = () => {
        const { xAxisKey, data } = this.props;
        const numDataPoints = data.length;
        // resize the font based on the number of datapoints (until we can find a more elegant fix)
        let fontSize = 15;

        if (numDataPoints > 25) {
            fontSize = 11;
        } else if (numDataPoints > 23) {
            fontSize = 12;
        } else if (numDataPoints > 21) {
            fontSize = 13;
        } else if (numDataPoints > 18) {
            fontSize = 14;
        }

        return (
            <XAxis
                minTickGap={10}
                axisLine={false}
                tickLine={false}
                tick={{ fontSize }}
                stroke="rgba(255, 255, 255, 0.8)"
                dataKey={xAxisKey}
                domain={["dataMin", "dataMax"]}
                dy={4}
                padding={{ left: 30, right: 30 }}
            />
        );
    };

    _renderYAxis = () => {
        return (
            <YAxis
                axisLine={false}
                tickLine={false}
                stroke="rgba(255, 255, 255, 0.8)"
                tickCount={6}
                tickFormatter={this._kFormatter}
            />
        );
    }

    _renderHoverLabel = (props) => {
        const { labelFormater } = this.props;
        const { barSelected } = this.state;
        const { x, y, width, height, value, index, labelKey } = props;
        const middleWidth = x + width / 2;

        return barSelected && (index === barSelected.index) && (labelKey === barSelected.key)
            ? (
                <g>
                    <line x1={middleWidth} y1={labelHeight} x2={middleWidth} y2={y + height} stroke="#57A0EA" />
                    <text x={middleWidth} y={labelHeight/2} fill="#fff" textAnchor="middle" dominantBaseline="middle">
                        {labelFormater ? labelFormater(barSelected) : value }
                    </text>
                </g>
            )
            : null;
    };

    _renderBars = () => {
        const { data, dataKeys, dataKeysStyle } = this.props;
        const { barSelected } = this.state;

        return _.map(dataKeys, (key) => {
            return (
                <Bar
                    key={`bar-${key}`}
                    dataKey={key}
                    name={key}
                    isAnimationActive={false}
                    minPointSize={5}
                    background={{ fill: "transparent" }}
                >
                    {data.map((entry, index) => {
                        const isHovered = barSelected && barSelected.key === key && barSelected.index === index;
                        const color = _.get(dataKeysStyle, `${key}.color`)
                            ? _.get(dataKeysStyle, `${key}.color`)
                            : "#fff";
                        const hoverColor = _.get(dataKeysStyle, `${key}.hoverColor`)
                            ? _.get(dataKeysStyle, `${key}.hoverColor`)
                            : color;

                        return (
                            <Cell
                                key={`bar-${entry}-${index}`}
                                onMouseOver={this._handleBarMouseOver(key, index, entry)}
                                onMouseOut={this._handleBarMouseOut}
                                cursor="pointer"
                                fill={isHovered ? hoverColor : color}
                                style={{ opacity: isHovered && (hoverColor === color) ? 0.6 : 1 }}
                            />
                        );
                    })}
                    <LabelList
                        dataKey={key}
                        labelKey={key}
                        content={this._renderHoverLabel}
                    />
                </Bar>
            );
        });
    };

    render() {
        const {
            "data-id": dataId,
            bgImage,
            loading,
            loadingMessage,
            errorMessage,
            title,
            legend,
            rockerButton,
            data,
            chartHeight,
            chartWidth,
            onGroupSelectionChange,
        } = this.props;
        const heroStyles = { backgroundImage: bgImage ? `url("${bgImage}")` : null };

        return (
            <div data-id={dataId} className="hero-chart" style={heroStyles}>
                {
                    errorMessage &&
                    <div className="hero-chart__error">
                        <Icon iconName="cogs" iconSize={iconSizes.XXL} />
                        <div className="hero-chart__error-text">{errorMessage}</div>
                    </div>
                }
                {
                    !errorMessage && loading &&
                    <PageSpinner className="hero-chart__loader" show>{loadingMessage}</PageSpinner>
                }
                {
                    !errorMessage && !loading && [
                        <ChartTitle key="title" className="hero-chart__title" title={title} />,
                        <Padding key="legend" vertical="sm">
                            <Legend
                                alignment={legendAlignments.CENTER}
                                boxAlignment={boxAlignments.CENTER}
                                data={legend}
                            />
                        </Padding>,
                        <ResponsiveContainer
                            key="multi-bar-chart-container"
                            className="hero-chart__top-chart"
                            height={chartHeight}
                            width="100%"
                        >
                            <BarChart
                                data-id={`${dataId}-multi-bar-chart`}
                                data={data}
                                height={chartHeight}
                                width={chartWidth}
                                margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
                                barGap="2"
                                barCategoryGap="30%"
                                margin={{ top: labelHeight }}
                            >
                                <CartesianGrid vertical={false} stroke="rgba(255, 255, 255, 0.4)" />

                                {/* tooltip is the only way for us to get any feedback when mousing over a group */}
                                <Tooltip
                                    cursor={false}
                                    content={<SectionHoverHandler onChange={onGroupSelectionChange} />} />
                                {this._renderYAxis()}
                                {this._renderXAxis()}
                                {this._renderBars()}

                            </BarChart>
                        </ResponsiveContainer>
                    ]
                }
                {rockerButton && <div className="hero-chart__rocker">{rockerButton}</div>}
            </div>
        );
    }
}
