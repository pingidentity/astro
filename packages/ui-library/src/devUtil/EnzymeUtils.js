import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
