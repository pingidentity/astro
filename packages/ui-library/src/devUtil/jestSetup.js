
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
