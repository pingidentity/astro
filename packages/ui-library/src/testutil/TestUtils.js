/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
