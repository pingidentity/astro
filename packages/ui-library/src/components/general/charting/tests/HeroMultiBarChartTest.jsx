import React from "react";
import ReactTestUtils from "react-dom/test-utils";
import { mount } from "enzyme";
import TestUtils from "../../../../testutil/TestUtils";
import HeroMultiBarChart from "../HeroMultiBarChart";
import _ from "underscore";
import { BarChart } from "recharts";
import { mountSnapshotDataIds } from "../../../../devUtil/EnzymeUtils";

jest.dontMock("recharts");
jest.dontMock("../HeroMultiBarChart");
jest.dontMock("../../../forms/RockerButton");
jest.useFakeTimers();

describe("HeroMultiBarChart", function () {
    const Wrapper = TestUtils.UpdatePropsWrapper;

    const chartData = [
        { id: "1/1", yups: 111, nopes: 222 },
        { id: "1/2", yups: 11111, nopes: 22222 },
        { id: "1/3", yups: 111111, nopes: 222222 },
    ];

    const defaultProps = {
        "data-id": "custom-chart",
        bgImage: "http://server.com/my/image/img.jpg",
        title: "Totals for Today",
        legend: [
            {
                label: "MFA",
                value: "1000",
                textColor: "#fff",
            }
        ],
        data: chartData,
        xAxisKey: "id",
        dataKeys: ["yups", "nopes"],
        dataKeysStyle: { nopes: { color: "#e34234", hoverColor: "#e34255" } },
        rockerButton: <div>Rocker button</div>,

    };

    const getComponent = props => {
        const componentProps = _.defaults(props || {}, defaultProps);
        return ReactTestUtils.renderIntoDocument(
            <HeroMultiBarChart {...componentProps} />
        );
    };

    const getMountedComponent = props => {
        const componentProps = _.defaults(props || {}, defaultProps);
        return mount(<HeroMultiBarChart {...componentProps} />);
    };

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <HeroMultiBarChart
                {...defaultProps}
            />
        );
    });

    it("renders with the default data-id", function () {
        const component = getComponent();
        const container = TestUtils.findRenderedDOMNodeWithDataId(component, defaultProps["data-id"]);

        expect(container).toBeTruthy();
    });

    it("renders the strings", function () {
        const component = getComponent();

        const title = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__title");
        expect(title.textContent).toBe(defaultProps.title);

        const legendInfo = TestUtils.findRenderedDOMNodeWithClass(component, "legend__item-info");
        expect(legendInfo.textContent).toBe(defaultProps.legend[0].label);

        const legendValue = TestUtils.findRenderedDOMNodeWithClass(component, "legend__item-value");
        expect(legendValue.textContent).toBe(defaultProps.legend[0].value);


    });

    it("renders the bg image", function () {
        const component = ReactTestUtils.renderIntoDocument(
            <Wrapper
                type={HeroMultiBarChart}
                {...defaultProps}
            />
        );

        const container = TestUtils.findRenderedDOMNodeWithDataId(component, defaultProps["data-id"]);

        expect(container.style.backgroundImage).toBe(`url(${defaultProps.bgImage})`);

        component._setProps({ bgImage: null });

        expect(container.style.backgroundImage).toBe("");
    });

    it("renders the x-axis", function () {
        const wrapper = getMountedComponent();
        const topChart = wrapper.find(BarChart).get(0);
        const xAxis = topChart.props.children[2];

        expect(xAxis.props.dataKey).toBe("id");
    });

    it("renders the x-axis in the various font sizes", function () {
        const fontSizeData = [
            { size: 15, threshold: 1 },
            { size: 14, threshold: 19 },
            { size: 13, threshold: 22 },
            { size: 12, threshold: 24 },
            { size: 11, threshold: 26 },
        ];

        fontSizeData.forEach((item) => {
            let customChartData = [];

            for (let i = 0; i < item.threshold; i += 1) {
                customChartData.push({ id: i, yups: 1, nopes: 1 });
            }

            const wrapper = getMountedComponent({ data: customChartData });
            const topChart = wrapper.find(BarChart).get(0);
            const xAxis = topChart.props.children[2];
            expect(xAxis.props.tick.fontSize).toEqual(item.size);
        });
    });

    it("renders the multi bars", function () {
        const wrapper = getMountedComponent();

        const barChart = wrapper.find(BarChart).at(0).instance();
        expect(barChart.props["data-id"]).toBe(`${defaultProps["data-id"]}-multi-bar-chart`);
        expect(barChart.props.data.length).toBe(chartData.length);
    });

    it("does not render the error or spinner by default", function () {
        const component = getComponent();

        const errorMessage = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__error");
        expect(errorMessage).toBeFalsy();

        const spinner = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__loader");
        expect(spinner).toBeFalsy();
    });

    it("renders the error when provided", function () {
        const errorMessageText = "Uh oh.";
        const component = getComponent({ errorMessage: errorMessageText });

        const errorMessage = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__error");
        expect(errorMessage).toBeTruthy();
        expect(errorMessage.textContent).toBe(errorMessageText);

        const title = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__title");
        expect(title).toBeFalsy();

        const barChart = TestUtils.findRenderedDOMNodeWithClass(
            component,
            `${defaultProps["data-id"]}-multi-bar-chart`
        );
        expect(barChart).toBeFalsy();

        const rockerButton = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__rocker");
        expect(rockerButton).toBeTruthy();
    });

    it("renders only the spinner when loading", function () {
        const loaderMessageText = "Wait!";
        const component = getComponent({
            loading: true,
            loadingMessage: loaderMessageText,
        });

        const spinnerContainer = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__loader");
        expect(spinnerContainer).toBeTruthy();

        const spinnerText = TestUtils.findRenderedDOMNodeWithClass(spinnerContainer, "page-loader__text");
        expect(spinnerText.textContent).toBe(loaderMessageText);

        const title = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__title");
        expect(title).toBeFalsy();

        const barChart = TestUtils.findRenderedDOMNodeWithClass(
            component,
            `${defaultProps["data-id"]}-multi-bar-chart`
        );
        expect(barChart).toBeFalsy();

        const rockerButton = TestUtils.findRenderedDOMNodeWithClass(component, "hero-chart__rocker");
        expect(rockerButton).toBeTruthy();
    });

    it("sets proper label on mouse over", function () {
        const component = getComponent();
        const barSelected = {
            key: "yups",
            index: 0,
            entry: { id: "1/1", yups: 111, nopes: 222 }
        };

        component._handleBarMouseOver(barSelected.key, barSelected.index, barSelected.entry)();
        component._renderHoverLabel({
            index: barSelected.index,
            labelKey: barSelected.key,
            value: barSelected.entry[barSelected.key]
        });
        expect(component.state.barSelected).toEqual(barSelected);
    });

    it("sets the selected element on mouse over with labelFormater", function () {
        const component = getComponent({ labelFormater: (value) => value });
        const barSelected = {
            key: "yups",
            index: 1,
            entry: chartData[1]
        };

        component._handleBarMouseOver(barSelected.key, barSelected.index, barSelected.entry)();
        component._renderBars();
        component._renderHoverLabel({
            index: barSelected.index,
            labelKey: barSelected.key,
            value: barSelected.entry[barSelected.key],
        });
        expect(component.state.barSelected).toEqual(barSelected);
    });

    it("clears the selected element on mouse out", function () {
        const component = getComponent();
        const barSelected = {
            key: "yups",
            index: 0,
            entry: chartData[0]
        };

        component._handleBarMouseOver(barSelected.key, barSelected.index, barSelected.entry)();
        expect(component.state.barSelected).toEqual(barSelected);

        component._handleBarMouseOut();
        component._renderHoverLabel({
            index: barSelected.index,
            labelKey: barSelected.key,
            value: barSelected.entry[barSelected.key],
        });
        expect(component.state.barSelected).toEqual(null);
    });

    it("_kFormatter format to proper format", function () {
        const component = getComponent();

        expect(component._kFormatter(111)).toEqual(111);
        expect(component._kFormatter(22222)).toEqual("22.2K");
    });
});
