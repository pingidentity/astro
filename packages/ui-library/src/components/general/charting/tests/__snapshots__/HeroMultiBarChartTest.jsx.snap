// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroMultiBarChart data-id's don't change 1`] = `
Array [
  "custom-chart",
  "padding-component",
]
`;
